src/forward/arweave-fetcher.go

```diff
@@ -0,0 +1,224 @@
+	self.db = db
+	return self
+}
+
+func (self *ArweaveFetcher) WithMonitor(monitor monitoring.Monitor) *ArweaveFetcher {
+	self.monitor = monitor
+	return self
+}
+
+func (self *ArweaveFetcher) WithInputChannel(input chan uint64) *ArweaveFetcher {
+	self.input = input
+	return self
+}
+
+func (self *ArweaveFetcher) run() (err error) {
+	for height := range self.input {
+		// Cache Contract id -> interaction sort key
+		lastSortKeys := make(map[string]string)
+
+		isFirstBatch := true
+
+		// Fetch interactions in batches, offset is the batch number
+		for offset := 0; ; offset++ {
+			// Fetch interactions in batches
+			var interactions []*model.Interaction
+
+			err = self.db.WithContext(self.Ctx).
+				Transaction(func(tx *gorm.DB) (err error) {
+					// Get a batch of L1 interactions
+					err = self.db.Table(model.TableInteraction).
+						Where("block_height = ?", height).
+						Where("source=?", "arweave").
+						Limit(self.Config.Forwarder.FetcherBatchSize).
+						Offset(offset * self.Config.Forwarder.FetcherBatchSize).
+
+						// FIXME: -----------------> IS THIS ORDERING CORRECT? <-----------------
+						// This is the order DRE gets L1 interactions
+						Order("sort_key ASC").
+						Find(&interactions).
+						Error
+					if err != nil {
+						return
+					}
+					if len(interactions) == 0 {
+						return
+					}
+
+					// Update last_sort_key for each interaction in the database
+					// As a optimization lastSortKeys are cached in memory
+					lastSortKeys, err = self.updateLastSortKey(tx, interactions, height, lastSortKeys)
+					if err != nil {
+						return
+					}
+
+					// Update sync height
+					return self.updateSyncedHeight(tx, height)
+				})
+			if err != nil {
+				self.Log.WithError(err).Error("Failed to fetch interactions from DB")
+				return
+			}
+
+			if len(interactions) == 0 && offset != 0 {
+				// Edge case: num of interactions is a multiple of batch size
+				payload := &Payload{First: false, Last: true, Interaction: nil}
+				self.Output <- payload
+			} else {
+				isLastBatch := len(interactions) < self.Config.Forwarder.FetcherBatchSize
+				for i, interaction := range interactions {
+					payload := &Payload{
+						First:       isFirstBatch && i == 0,
+						Last:        isLastBatch && i == len(interactions)-1,
+						Interaction: interaction,
+					}
+
+					// NOTE: Quit only when the whole batch is processed
+					// That's why we're not waiting for closing of this task
+					self.Output <- payload
+				}
+			}
+
+			isFirstBatch = false
+		}
+	}
+	return
+}
+
+func (self *ArweaveFetcher) updateSyncedHeight(tx *gorm.DB, height uint64) (err error) {
+	var state model.State
+	err = tx.WithContext(self.Ctx).
+		Where("name = ?", model.SyncedComponentForwarder).
+		First(&state).
+		Error
+	if err != nil {
+		self.Log.WithError(err).Error("Failed to get state")
+		return err
+	}
+
+	if state.FinishedBlockHeight < height {
+		state.FinishedBlockHeight = height
```

Comment:
not sure I follow - how do we know at this point that we've processed ALL the L1 interactions for the given `height` and that `FinishedBlockHeight` can be safely set?

Shouldn't this be run (maybe it already is - but I don't see that in code) only for interactions from the last batch for the given `height`? which can be easilly verified (number of returned interactions < limit)